### DISCUSSION TITLE
Should we redefine our concept of "keyword independence"?

### DISCUSSION THREAD

**Discussion by jdesrosiers** [json-schema-org/community#204] [Specification]
The various discussions lately about the architecture of JSON Schema got me thinking about our principle of "keyword independence", why we value it, and why we seem to constantly violate it.

For those who aren't familiar, the keyword independence principle is the idea that keyword behaviors should be self contained. You shouldn't need information from other keywords in order to evaluate it. For example, `minimum` is ok, but `additionalProperties` is a violation because it depends on `properties` and `patternProperties`.

Keywords that break this principle can have negative properties for schema design, but the overwhelming consensus is that the downsides in most cases are minor and better than the alternatives if there are any.

For me, the big win from keyword independence is the simplicity of the processing model it allows. Keywords can be evaluated in any order. No state needs to be maintained and passed around. This means you can evaluate keywords in parallel if the implementation's programming language supports it, which could mean big performance improvements in many cases.

It turns out that as long as the information we need from the depended on keywords can be determined statically (without evaluating the keyword), we can have keyword dependence without losing all those nice properties in the previous paragraph. For example, `additionalProperties` can look at the keys in the `properties` and `patternProperties` objects and have all the information it needs to evaluate against an instance independently. This can even be done in a compile step.

So, if we consider keyword independence to be whether a keyword can be evaluated independently of the evaluation of other keywords, there are really only a few problematic keywords. These include the `then` and `else` keywords that depend on the evaluation result of `if` and the `unevaluatedProperties` and `unevaluatedItems` keywords that depend on the evaluation of sub-schemas. Technically, these violating keywords can be implemented to be independent, but it would result in evaluating some parts of the schema more than once in some cases.

I think switching to this definition of keyword independence gives us a more clear and compelling reason for the principle than we had before. It also allows us to stop considering certain keywords in violation that haven't really been problematic. With a clearly defined reason for the principle, we can focus on trying to improve the few keywords that are in violation or explain why we think it's worth the violation.

**Comment by gregsdennis** [comment]
> `additionalProperties` can look at the keys in the `properties` and `patternProperties` objects and have all the information it needs to evaluate against an instance independently.

This then requires the keyword to have knowledge of the schema in which they reside.  While your concept grants that `additionalProperties` doesn't need the _evaluation result_ of these other keywords, it does depend on their presence (insofar as to say their absence is like they're empty).

> This means you can evaluate keywords in parallel if the implementation's programming language supports it, which could mean big performance improvements in many cases.

While you may not be fully processing the dependent keywords in your static analysis, you still need to minimally consider them multiple times, which can subvert gains you get from parallel processing.

My opinion is that we should have keyword independence as a goal, but not a hardfast requirement.  Supposing that a keyword can be designed to work in such a way that allows it to be independent of other keywords, it should be.  However we should recognize that sometimes keywords need the results from other keywords, and we need to be okay with this.

Additionally, you still get the benefit of parallel processing if you group by dependency depth.  For example, `additionalProperties` depends on `properties` and `patternProperties`, but those two are independent, and `unevaluatedProperties` depends on all three.  So `properties` and `patternProperties` have a dependency depth of 0,  `additionalProperties` has a dependency depth of 1, and `unevaluatedDependencies` has a dependency depth of 2 (and that's really as deep as it goes).  Grouped like this, all 0-depth keywords can be processed in parallel, followed by all 1-depth keywords, etc.

**Reply by jdesrosiers** [reply]
> This then requires the keyword to have knowledge of the schema in which they reside.

Yes, the point of this redefinition is to expand the context that the keyword as access to from just the keyword value to the schema the keyword is in. The input to the keyword handler would include a schema and a pointer to the keyword in the schema. Getting information about other keywords is a matter of manipulating the pointer. Things like looking at a sibling keyword is a trivial operation.

> While you may not be fully processing the dependent keywords in your static analysis, you still need to minimally consider them multiple times, which can subvert gains you get from parallel processing.

It depends on the level of static analysis necessary. That would have to be taken into account when designing keywords. Generally I think it's unlikely that it would outweigh the ability to parallelize, but I don't have numbers to prove that. In any case, that's not an issue with implementations that compile schemas because that minimal amount of extra work happens in the compile step and doesn't affect validation cost.

> Additionally, you still get the benefit of parallel processing if you group by dependency depth.

Yes, but one of the benefits of the redefinition is to simplify the processing model. This requires adding a concept of dependency depth to the model which isn't necessary otherwise.

> Supposing that a keyword can be designed to work in such a way that allows it to be independent of other keywords, it should be.

I agree with this, but as one factor to consider. There's always a way to define keywords to be independent, but there are other factors that might lead us to choose otherwise. I think fully independent is preferred and if that doesn't make sense, then schema independent is preferred. We may still decide to add keywords that aren't even schema independent (like `unevaluated*`) if we think there is sufficient value in doing so, but we should be cautious in these cases because they will be more difficult to implement.

Therefore, maybe it makes sense to add this concept rather than replace the old one. There's still value to the current definition, it's just not as strong as the value of the proposed new definition. It's possible to have both.

**Comment by gregsdennis** [comment]
Related to https://github.com/json-schema-org/json-schema-spec/issues/701

**Reply by handrews** [reply]
OK I'm finally getting to this now that I've published the keyword behaviors ideas (#220).

@jdesrosiers I think this nicely illustrates the value of your early feedback on the behaviors regarding deeper underlying principles (compared to the level on which I focused) that can help us conceptualize JSON Schema.

What I get from this discussion is not so much a redefinition of "keyword independence" (because a static dependency is still a dependency) but an examination of the deeper distinction between static and dynamic/runtime processing.  Which, having read this discussion, seems like a distinction worth exploring more deeply.

_(This next bit is me walking through how we ended up with a runtime-centric notion of keyword dependencies, because in the spirit of "the map is not the territory", I want to show how my "map" imposed on the "territory" of the specification, so we can select a better map — prior to Jason's feedback about deeper principles in a slack conversation, I would have been more inclined to defend the current setup.  The intended audience is largely folks following the discussion, so I'm covering things that I know that Jason and Greg already know.)_

This distinction shows up in the [keyword behaviors](https://drive.google.com/drive/folders/1YOklA0UbS0-XU8aWvY9UmCLoOEzoffXF?usp=sharing) primarily in the [identifier (D)](https://docs.google.com/presentation/d/1piOtNwK5_CgmDZSneLRE4ZxHrKruJHwy9Aa9gMHfe7w/edit#slide=id.g1400c909ee2_0_227) and [lookup (K)](https://docs.google.com/presentation/d/1piOtNwK5_CgmDZSneLRE4ZxHrKruJHwy9Aa9gMHfe7w/edit#slide=id.g1400c909ee2_0_647) behaviors, each of which is split into lexical/static (x) and dynamic/runtime (y) variations.  On the other hand, we currently specify keyword interactions in terms of [annotation (N)](https://docs.google.com/presentation/d/17b1o_nIm2KkYgdMfoAlcHss4FTFEwISIc3oVzCEN-h8/edit#slide=id.g14275c8723d_0_148) and [consumer (U)](https://docs.google.com/presentation/d/17b1o_nIm2KkYgdMfoAlcHss4FTFEwISIc3oVzCEN-h8/edit#slide=id.gfdf17d4ee3_0_945) behaviors, which have no such distinction.  Which makes sense at the keyword behavior level because N behavior is about attaching things to instances.  So, me having my N-behavior blinders on meant that I never really thought about static options for keyword communication, even though I was aware that you could calculate `additionalProperties` behavior without an instance.

And even though I've observed that there's otherwise a symmetry between [K-D interactions](https://docs.google.com/presentation/d/1piOtNwK5_CgmDZSneLRE4ZxHrKruJHwy9Aa9gMHfe7w/edit#slide=id.g1400c909ee2_0_819) (finding information in parent scopes) and [U-N interactions](https://docs.google.com/presentation/d/17b1o_nIm2KkYgdMfoAlcHss4FTFEwISIc3oVzCEN-h8/edit#slide=id.gfdf17d4ee3_0_1005) (reading information from child scopes) that static/runtime distinction is not present.

@gregsdennis noted that even with static analysis there is still some duplication of effort, which is definitely true if you do the static analysis at runtime.  But we already have a concept of statically analyzing a schema at [load time](https://docs.google.com/presentation/d/192uw7T2DftoI4r3Zz1c1acMN2_BwYdZKT3gbuEaG_Yc/edit#slide=id.gfdd23b4381_0_366).  This step is [mentioned](https://www.ietf.org/archive/id/draft-bhutton-json-schema-01.html#name-loading-a-schema) in the spec but not required to be implemented separately.  However, resolving certain kinds of references is tricky if you don't have a step like that.

In addition to indexing schemas by URI by finding and resolving`$id` and `$anchor` (and in some implementations, also indexing by JSON Pointer fragment URIs as well), a static pass could also do things like resolving `$ref` values against base URIs so that the full URI is available at runtime.  A static analysis to support `additionalProperties`, `items`, and `contains` could also happen at that time, with the results available at runtime.

I'm not saying that it _has_ to happen then — @jdesrosiers's access to the schema object at runtime of course also  works.

But I do think that if we get more intentional and clear about static vs runtime behavior this becomes more straightforward, and better integrated with other things that area already happening in JSON Schema whether they are directly mandated by the specification or not.

I would object to redefining "independent" to mean that `additionalProperties` is independent, because it's really not.  But I do think that talking about static vs runtime dependencies would be very helpful, and would fit into similar concerns regarding other static (`$id`, `$anchor`, `$defs`, `$ref`) vs dynamic (`$dynamicAnchor`, `$dynamicRef`) keywords and their costs and benefits.

It would also probably help reduce the feeling that dynamic behaviors are weird and exceptional.  And that's really a lot of what I was trying to do by defining dependencies in terms of annotations:  normalize what was necessary, particularly to implement `unevaluated*`, and tie it to other desirable behaviors.  But the lack of distinction between static and dynamic in this context meant that the more complex dynamic requirements dominated when they do not really need to.

**Reply by gregsdennis** [reply]
I recall that we actually did have quite a bit of discussion around the "can't be determined until evaluation time" topic when we were defining `$recursive*`.  My recollection of it mostly lines up with this, but it's worth mentioning that we've been here before.

**Reply by handrews** [reply]
@gregsdennis good point.  We definitely talked about evaluation time/runtime behavior regarding both `$recursive*` and `unevaluated*`, but it was kind of piecemeal.  I don't recall really developing a good theory of it that tied everything together (and if we did, I forgot it at some point).

**Reply by jdesrosiers** [reply]
@handrews Thanks for the thoughtful analysis. I agree with almost everything.

> I would object to redefining "independent" to mean that `additionalProperties` is independent, because it's really not.

Agreed. I've changed my mind about redefining the term. However, I still think we need to revisit how we present it in the spec. I think the spec can stop making big a deal about static keyword independence since it isn't a significant problem and we clearly aren't committed to not allowing it or removing keywords that violate it. At best, we need one section that says we prefer keyword independence whenever possible, but that sounds more like documentation content rather than spec content.

> It would also probably help reduce the feeling that dynamic behaviors are weird and exceptional.

This is the one place where I disagree slightly. I think dynamic behaviors should be considered weird and exceptional. They are harder to implement, harder to reason about, and often have poor performance. I think dynamic references and the unevaluated keywords bring value, but I don't think we should generally encourage more keywords with runtime dependencies. I don't mean that they shouldn't be included in the model, just that it should be discouraged.

**Reply by handrews** [reply]
@jdesrosiers

> I think the spec can stop making big a deal about static keyword independence since it isn't a significant problem and we clearly aren't committed to not allowing it or removing keywords that violate it. At best, we need one section that says we prefer keyword independence whenever possible, but that sounds more like documentation content rather than spec content.

Yes, this makes sense to me.

> This is the one place where I disagree slightly. I think dynamic behaviors should be considered weird and exceptional. They are harder to implement, harder to reason about, and often have poor performance.... I don't mean that they shouldn't be included in the model, just that it should be discouraged.

I'd like to offer a different perspective on this, which is that _dynamic behaviors involve certain trade-offs that should be considered._  This is a characteristic they share, albeit with different details, with other sorts of keywords such as `uniqueItems`.  

Keywords that assert uniqueness across a potentially arbitrarily large set of data impose substantial memory requirements, which can impact performance as well.  A decision whether to add (or support) such keywords should consider these tradeoffs.

Similarly, keywords that reference remote resources, whether those resources are schemas or some other external data (e.g. Greg's `data` keyword, which can access arbitrary external resources in addition to the instance data) also involve tradeoffs including environmental requirements, error modes, and security concerns.  Another set of tradeoffs to consider within the space of allowable keyword behaviors (which we are still defining).

As one last bit of context, we do need to remember to think of the larger ecosystem, where keywords are plugins.  We don't exactly have a huge backlog of dynamic keywords to add to the actual standard (I can't think of any, TBH).  But we need to give people ways to think about keyword design, and "don't do things like this thing that we did" is unlikely to be effective.  Ensuring that the cost of such keywords is contained in the plugins, with predictable impact on the core implementation, will be  more robust.

------

So when it comes to dynamic runtime behavior, let's look at your stated concerns.  The rest of this comment is me trying to think through how we got here with dynamic behavior, and where we might go.  Perhaps the most significant part is near the beginning, looking at code size and complexity in jschon.  Although that might not be directly relevant to implementations that take a more "compiled" approach to schema evaluation.

> They are harder to implement

I think this is primarily true when making the shift from the draft-07 paradigm to the modern paradigm.  They can be quite hard to implement in an architecture that was not designed with them in mind.  But looking at 2019-09-and-later-only implementations, the code for these keywords is quite straightforward.

In jschon, `additionalProperties` and `unevaluatedProperties` are each implemented in 33 lines of code.  `$anchor` and `$dynamicAnchor` are each 13 lines of code.  `if`, `then`, and `else` are tiny (6, 9, and 9 lines, respectively). 
 `unevaluatedItems` (43 lines) is longer than `items` (25 lines), but most of that (13 lines) is because it depends on four keywords while `items` depends on only one.  The biggest divergence is `$ref` (22 lines) and `$dynamicRef` (54 lines), but `$dynamicRef` is uniquely complex.  Your current proposal to change it would reduce that code size somewhat.  My proposed alternative (which I do not plan to push for) would reduce it even more as part of it could just be handed off to existing `$ref` code.

It's definitely true that `$dynamicRef` and `unevaluatedItems` are the keywords that take the most lines of code to implement in jschon, but with `additionalProperties` at 33 lines, 54 lines (and likely to shrink) isn't problematically huge.

Including these keywords in the model, whether there are many or few such keywords, means normalizing the underlying support as part of the architecture.  That's what determines whether a keyword is hard or easy to implement:  If it requires something outside of the existing architecture, it's hard.  But going forward, the support will be baked-in, and that's why the code in jschon is straightforward: it was built assuming these features to begin with.  I'm sure the actual amount of code involved is substantially larger, but most of it is not keyword-specific.

Regardless of whether the way I defined keyword behaviors is the correct granularity or level of abstraction, it's important to shift to defining those underlying requirements and ensuring that implementers understand the division between generic and keyword-specific functionality.

> harder to reason about

Of your objections, this is the one that I find the most compelling.  I do think that part of this is still about adjusting assumptions from draft-07 and earlier, but not all of it.  It also varies across the different keywords.

In terms of usage, I don't think anyone has any problem with `if`/`then`/`else`.  It's hard for use cases like code generation, but no more so than `not`.  Really, the runtime behavior here is only because we made it three separate keywords.  If it had been one keyword with three subschemas, it would be no more complex than `oneOf`.

With `unevaluated*`, most people to seem to get it, but we have seen some definite confusion.  I think some of this has to do with communication, and the core team's own journey to figure this out (e.g. the "unevaluated" vs "un-successfully-evaluated" discussion, the questions of whether keyword failure or overall schema failure impacts annotations, etc.)  I think it's possible to present these in ways that are more intuitive, but as things stand there is a valid concern.

For `$dynamicRef` and `$dynamicAnchor`, even I have trouble with these keywords despite creating them and writing the spec language.  The changes under discussion in spec issue 1140 will make reasoning about these easier, but fundamentally they're trying to do a thing that's challenging to conceptualize in the first place.  This is actually why I proposed a `$dynamicRef` that has two values, splitting the two steps involved in resolving it:  each step is much easier to reason about than the whole thing at once.  (at least for me- obviously you did not find it so).

I don't have a neat conclusion for this concern, because it's quite valid.  But I do think that some of it is growing pains and us not understanding how to create and communicate these keywords at first.  There's a parallel here with `id`/`$id`, which was historically viewed as very complex and difficult.  But after splitting `$anchor` out and introducing the concept of schema resources (as opposed to some schema objects just kind of randomly having `$id` for unclear reasons), it's much more straightforward without losing any of the functionality.

> often have poor performance

This one I do want to dispute, at least in part.  I suspect (but don't know because I lack experience with it), that when it comes to "compiling" schemas for better performance, these keywords (except perhaps `if`) present unique challenges and impose unique costs.  It's probably worth a deeper examination by someone who knows what they're talking about.  Which is not me.

 But setting aside schema compilation, poor performance compared to what?  Performance has to be evaluated in the context of functionality, and also whether there is a performance impact when the keyword is not used.  For example, `uniqueItems` can have horrible performance, but only if used, and the performance problem is inherent in the functionality.

`if`/`then`/`else` do not have poor performance (I don't think you had these keywords in mind, I'm just including them for completeness).

`unevaluated*` requires storing and examining the annotations involved.  As I mentioned on a call a few weeks ago, I have ideas on mitigating this (and annotation collection costs in general).  Specifically around ensuring that annotations are only collected if they will be needed.  This would substantially reduce the cost imposed by supporting these keywords when they are not being used.  When in use, the cost of examining the annotations is unavoidable:  you can't have this feature without examining that behavior, and of all the dynamic features added this one has the longest and most persistent history of being wanted.  Often in the form of people assuming that `additionalProperties` already did this and complaining that the implementation was "broken."

`$dynamic*` keywords are better about not imposing costs when they are not in use.  There's definitely some cost here (with might get better in the next release), but as with `unevaluated*` there is no viable workaround if we want to have this sort of thing at all.  Unlike `unevaluated*`, these keywords will probably always be relatively rare.  They feel like reasonable tradeoffs given the possibilities they open up (e.g. custom meta-schemas are actually practical now, when they really kind of were not before).

**Reply by jdesrosiers** [reply]
@handrews Your perspective is valid. You're absolutely right that this is about tradeoffs.

> For me, the big win from keyword independence is the simplicity of the processing model it allows. Keywords can be evaluated in any order. No state needs to be maintained and passed around. This means you can evaluate keywords in parallel if the implementation's programming language supports it, which could mean big performance improvements in many cases.

This is what I think is lost with dynamic behaviors. So, the question is: is the benefit of dynamic behaviors worth that loss? Personally, I think it can be worth it in exceptional cases, but not in the general case. Depending on what you place more importance on, you can easily come out on the other side given the same tradeoff.

I don't want to get too into the rest of your analysis, but I'll address a couple points briefly.

Harder to implement doesn't necessarily mean a lot of code. I found dynamic references to be one of the most challenging parts of the spec to implement despite ultimately being a shocking small number of lines of code.

As for performance, I've found that simply maintaining state for dynamic references had a measurable effect on my implementation's performance. There were similar reports about ajv. Evaluating a dynamic reference isn't computationally expensive. It's maintaining state that seems to be the costly factor.

**Reply by handrews** [reply]
> Harder to implement doesn't necessarily mean a lot of code. I found dynamic references to be one of the most challenging parts of the spec to implement despite ultimately being a shocking small number of lines of code.

This is an excellent point.  You'll hear no disagreement from me that dynamic references are hard - witness me repeatedly getting things wrong when talking about them despite having come up with them.  With dynamic references, the question is whether they're hard because dynamic stuff is hard, or hard because they're extra-complicated.  I think it's because they're extra-complicated.  If they *only* had dynamic behavior, they would be fairly simple.  It's the combination with lexical (resource-scoped) behavior that makes them really hard to think about.

> As for performance, I've found that simply maintaining state for dynamic references had a measurable effect on my implementation's performance. There were similar reports about ajv. Evaluating a dynamic reference isn't computationally expensive. It's maintaining state that seems to be the costly factor.

That's unexpected to me.  Can you go into a little more detail?  Is the problem the computation cost of maintenance or is it less direct?

**Reply by jdesrosiers** [reply]
> With dynamic references, the question is whether they're hard because dynamic stuff is hard, or hard because they're extra-complicated.

They are extra complicated, but for me the extra-complicated bits weren't what made it hard to implement. It was thinking through how things behaved dynamically during evaluation and figuring out how to represent that. But that's more a "hard to reason about" thing than a "hard to implement" thing. So, maybe it's only hard to implement because it's hard to reason about.

> Can you go into a little more detail?

I never had the time to profile it in detail, but here's what I think the problem is. State can be difficult to manage because it's not just one global state. It can grow and shrink as you go down one branch and return to follow another branch. Rather than trying to figure out what state needs to be removed in which cases, I use immutable state objects. When I go down one path and it's state needs to be dropped, I abandon the copy of the state that went down that path and use the copy I had before I went down that path. I don't think I explained that very well. I hope it made sense. Anyway, every time I change the state I effectively make a copy because the state is immutable. All that object creation can get expensive. I'm sure I could make it work with a mutable state, but I expect that would be significantly more complicated. It would also make it impossible to implement parallelized evaluation, which is something I want to do at some point.

**Reply by handrews** [reply]
Thanks, that does make sense to me.  I'd actually love to put some real focus on performance for the next release (more on that in a longer reply I'm working on), although the "next release" is kind of overloaded with major questions right now.

Both parallelization and schema compilation are techniques that I also think have been under-considered (certainly by me) over the last several drafts.  These arguably fall under the broader umbrella of "performance" but each have their own concerns.

You mention several tradeoffs in that last sentence that are worth exploring in detail in terms of what the spec demands.  There are quite a few options here, including defining restricted subsets (whether in terms of keywords, behavior, principles, whatever) for certain well-understood tradeoffs.  @karenetheridge's [JSON::Schema::Tiny](https://metacpan.org/pod/JSON::Schema::Tiny) is an inspiration here.  @jdesrosiers I know your own implementation is modular, but I haven't looked into how you've broken up the functionality.

Part of what I wanted to do with vocabularies (and the reason `unevaluated*` got moved to their own vocabulary, and perhaps something similar could/should happen with `$dynamic*` although its use in meta-schemas makes that more complicated but not impossible) was move away from JSON Schema being everything to everyone.  You've raised some significant questions about the granularity of vocabularies vs keywords, but wherever that lands, giving implementers a clear way to express "you can do these things but not those things, and it's not some weird personal decision but a choice that _is intentionally accommodated by the spec_" is an option I think we should consider.

**Reply by gregsdennis** [reply]
> Rather than trying to figure out what state needs to be removed in which cases, I use immutable state objects. When I go down one path and it's state needs to be dropped, I abandon the copy of the state that went down that path and use the copy I had before I went down that path.

Mine did this as well, until I had a user suggest a mutable context object (my state).  Now my state object is built with individual stacks, and I have to manage it.  I'm not sure why it makes a significant difference because it's still tracking the same amount of data, but it seems to.

I need to profile mine as well.  Corvus (.net) claims to have a minimal allocation implementation, but I don't know if the test they showcase was specially selected or not.

**Reply by gregsdennis** [reply]
> maybe it's only hard to implement because it's hard to reason about.

This was it for me. Once I got my head around it, building it was straightforward.

**Reply by handrews** [reply]
> It was thinking through how things behaved dynamically during evaluation and figuring out how to represent that. But that's more a "hard to reason about" thing than a "hard to implement" thing.

@jdesrosiers @gregsdennis I realize it's very hard to do a hypothetical on something that you've had to work through in a very different way already, but does the following way of framing dynamic lookups/identifiers (using a hypothetical `$da` and `$dr` that don't do everything the current `$dynamic*` keywords do) feel easier to reason about?  Both of these keywords only have effect when the schema object that contains them is evaluated.

* `"$da": "x"` associates "x" with the current dynamic scope _if and only if_ "x" is _not_ present in any parent dynamic scope
* `"$dr": "x"` looks for the nearest parent scope with an "x" associated with it by `$da`, and resolves to the associated schema location

"schema location" here is the URI captured in standard output for any annotations or errors for that dynamic scope, currently as `absoluteSchemaLocation`, and in the future as just `schemaLocation`.

**Comment by handrews** [comment]
@jdesrosiers pulling this out to a top-level comment b/c the other thread has a good discussion going in a slightly different direction:

> So, the question is: is the benefit of dynamic behaviors worth that loss [of performance and the ability to do parallel evaluation]? Personally, I think it can be worth it in exceptional cases, but not in the general case.

But what does that "exceptional" vs "general" really mean in practice?

From a framework+modular keywords perspective, dynamic behaviors are either supported or not.  There are two main dynamic behaviors:  looking up an identifier in parent dynamic scopes, and reading annotations from child dynamic scopes.  When it comes to cost, there are three areas where cost is incurred:

1. framework code (meaning anything shared by multiple keywords, whether your implementation makes a clear division between "framework" and "keyword" or not)
2. standard keywords
3. 3rd-party extension keywords.

_[side note:  I'm going to ignore `if`/`then`/`else` because it's only the requirement that no more than one of `then` or `else` be evaluated that makes it non-parallelizable.  That requirement could easily be relaxed as long as annotations are deleted correctly for the unnecessary branch.]_

## Framework costs

Framework support for these behaviors is a one-time cost.  Unless we eliminate them entirely (and `unevaluated*` and `$dynamic*` with them), that cost is factored in.  I've always been concerned about the annotation cost, and filed #236 to address that.  As noted above, I did not expect parent lookups to be expensive, so I'm interested in what you have to say about that (since it's state-related, it might be relevant to annotations as well- identifiers and annotations basically only differ in direction of usage (parent to child or child to parent).

We can and should work to reduce this cost, but the only way to eliminate it is to eliminate those keywords, and I don't think you or anyone else are proposing that we do that (please correct me if I am wrong).

We should also work to understand how to make these keywords as parallelization-friendly as possible, while acknowledging that they do impose some serial execution requirements when they are used.  We should work to ensure that they do not reduce parallelism when they are not in use.

## Standard keyword costs

The area of biggest concern would be standard keyword costs, because most implementations will (correctly) feel obligated to support these, and each such keyword can add additional cost and impose additional serial execution requirements.  Part of why I filed #236 is to ensure that, as much as possible, standard keywords with dynamic behavior only incur (most of) their cost when they are used.

While there are coarse-grained optimizations that can be done now (e.g. don't collect annotations if `unevaluated*` are not being used and the caller did not ask for annotations to be collected), we can make that a lot more fine-grained.  I would hope that we can do something similar with `$dynamic*` if you have not already done that.

But as far as I know, there aren't any more dynamic keywords in high demand.  I certainly have no plans to propose any, and the ones we have all met much higher bars of demand and/or discussion before they were accepted than most anything else in JSON Schema.  I would hold other dynamic keywords to similar standards, and would support (and write if folks would like) an ADR making that high bar very clear.

***So if your concern about exceptional vs general case is primarily about standard keywords, I think I agree and I would like to help codify that.***

## 3rd-party extension keyword costs

This is where we should have a general discussion about what our obligation is regarding what people _could_ do with extension keywords.  My keyword behaviors proposal is intended to put some boundaries on that so that there's a limited possibility for extensions to kill JSON Schema understandability and performance and damage the overall reputation of the project.  Your counter-idea of doing this at a different abstraction level of principles would presumably accomplish the same goal (and I haven't tried to move the keyword behaviors stuff towards anything concrete because I honestly don't know which level will work out better).

Whether we're talking about behaviors or principles, separating those things into "framework costs" (whether there's a literal framework or not) that are tightly constrained by the spec, vs "keyword costs" that might vary substantially, limits some of the possible costs of 3rd-party keywords.  Which is obviously a more general concern than just dynamic behaviors, although dynamic behaviors are among the costs we most want to contain.

Beyond these containment strategies, I don't think there's much we can or should do.  We could exclude dynamic behaviors from the modular keyword "API" (in quotes b/c I don't mean literally specifying an API, more of a conceptual thing).  But that would be hard to enforce and might actually make people think about those behaviors more.  Plus, it would preclude valuable explorations of the tradeoffs.

## To summarize...

* The most intractible costs are framework costs, and we just have to manage those through #236 and similar discussions — _this is where specification work is relevant_
* To the extent that anything should be "discourage", I think that is best done with an ADR establishing a high bar for accepting new dynamic keywords into the standard JSON Schema Org-maintained vocabularies —_this sort of guidance doesn't belong in the spec, but does belong in a formal Org-published document of some sort_
* Since the ADR in the previous point should make the trade-offs clear and communicate that throwing a bunch of dynamic keyword proposals at the spec team is unlikely to be successful, I don't think anything beyond that should be done about potential 3rd-party dynamic keywords — _there's just no upside in making a big deal about it_

**Reply by jdesrosiers** [reply]
> But what does that "exceptional" vs "general" really mean in practice?
>
> From a framework+modular keywords perspective, dynamic behaviors are either supported or not. 

I haven't read the whole comment and I'm too tired to today, but I can explain what I meant here. If something is generally supported it should be part of the framework and have plugin support, etc. If it's not generally supported, it's something that would have to be built into the implementation. Effectively that would mean that we as JSON Schema Org can make exceptions, but there wouldn't be support for third parties to make vocab keywords with dynamic behaviors.

That's probably too extreme. What I had in mind was more like it being optional for implementations to support third-party vocab keywords with dynamic behaviors.

**Reply by handrews** [reply]
> If something is generally supported it should be part of the framework and have plugin support, etc.

> What I had in mind was more like it being optional for implementations to support third-party vocab keywords with dynamic behaviors.

This actually dovetails with some things related to your process proposal that I'm going to open a new discussion on, about framework vs features vs keywords.  I'll link it here when I do.  Feel free to continue here with anything that feels relevant.
